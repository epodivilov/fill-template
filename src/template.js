/**
 *
 * @typedef {Object} Node
 * @property {'text'|'variable'} type
 * @property {string} value
 *
 * @typedef {Object} Container
 * @property {'root'|'group'} type
 * @property {Array<Node|Container>} children
 */

/**
 *
 * @param {string} value
 * @returns {Container}
 */
export function parse(string) {
  /** @type {Container} */
  const root = {
    type: "root",
    children: [],
  };

  let stack = [root];
  let type = "text";
  let value = "";

  let idx = 0;
  while (idx < string.length) {
    const ch = string.charAt(idx);

    if (ch === "[" || ch === "]" || ch === "{" || ch === "}") {
      if (value.length > 0) {
        stack[0].children.push({ type, value });
      }
    }

    switch (ch) {
      case "[": {
        const group = { type: "group", children: [] };
        stack[0].children.push(group);
        stack.unshift(group);

        type = "text";
        value = "";
        break;
      }
      case "]": {
        stack.shift();

        type = "text";
        value = "";
        break;
      }
      case "{": {
        type = "variable";
        value = "";
        break;
      }
      case "}": {
        type = "text";
        value = "";
        break;
      }
      default: {
        value += ch;
      }
    }

    idx++;
  }

  if (value.length > 0) {
    stack[0].children.push({ type, value });
  }

  return root;
}

/**
 *
 * @param {Container} ast
 * @returns {string}
 */
export function compose(ast, values = {}) {
  let result = "";

  /** @type {Array<Node|Container>} */
  const nodes = [ast];

  while (nodes.length > 0) {
    const node = nodes.shift();

    if (node.type === "root" && Array.isArray(node.children)) {
      nodes.unshift(...node.children);
      continue;
    }

    if (node.type === "group" && Array.isArray(node.children)) {
      const hasUndefinedVariable = node.children.some(
        (it) => it.type === "variable" && values[it.value] == null
      );

      if (!hasUndefinedVariable) {
        nodes.unshift(...node.children);
      }

      continue;
    }

    if (node.type === "text") {
      result += node.value;
      continue;
    }

    if (node.type === "variable" && values[node.value]) {
      result += values[node.value];
    }
  }

  return result;
}

/**
 * Fill template with variables
 * @param {string} string
 * @param {Record<string,string>} values
 */
export function fillTemplate(string, values = {}) {
  const ast = parse(string);

  return compose(ast, values);
}
