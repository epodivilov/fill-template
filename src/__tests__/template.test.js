import { fillTemplate, parse, compose } from "../template";

describe("fillTemplate", () => {
  it.each`
    input                                                          | params                               | output
    ${"/root[/{tenant}][/object-{id}]"}                            | ${{}}                                | ${"/root"}
    ${"/root[/{tenant}][/object-{id}]"}                            | ${{ tenant: "moskva" }}              | ${"/root/moskva"}
    ${"/root[/{tenant}][/object-{id}]"}                            | ${{ id: "12345" }}                   | ${"/root/object-12345"}
    ${"/root[/{tenant}][/object-{id}]"}                            | ${{ tenant: "moskva", id: "12345" }} | ${"/root/moskva/object-12345"}
    ${"Hello[, {name}]"}                                           | ${{}}                                | ${"Hello"}
    ${"Hello[, {name}]"}                                           | ${{ name: "World" }}                 | ${"Hello, World"}
    ${"[{greatening}, ]World"}                                     | ${{}}                                | ${"World"}
    ${"[{greatening}, ]World"}                                     | ${{ greatening: "Hello" }}           | ${"Hello, World"}
    ${"Example[ with {group} groups[ and {variable} variables]]."} | ${{}}                                | ${"Example."}
    ${"Example[ with {group} groups[ and {variable} variables]]."} | ${{ variable: 2 }}                   | ${"Example."}
    ${"Example[ with {group} groups[ and {variable} variables]]."} | ${{ group: 2 }}                      | ${"Example with 2 groups."}
    ${"Example[ with {group} groups[ and {variable} variables]]."} | ${{ group: 2, variable: 2 }}         | ${"Example with 2 groups and 2 variables."}
  `(
    "fill template '$input' with params $params should be equal '$output'",
    ({ input, params, output }) => {
      expect.hasAssertions();
      expect.assertions(1);

      expect(fillTemplate(input, params)).toBe(output);
    }
  );
});

describe("parse", () => {
  it("simple string", () => {
    expect(parse("Test string")).toMatchSnapshot();
  });

  it("string with variable", () => {
    expect(parse("Test string with {var}")).toMatchSnapshot();
  });

  it("string with group", () => {
    expect(parse("Test string[ with group]")).toMatchSnapshot();
    expect(parse("[Test string ]with group")).toMatchSnapshot();
  });

  it("string with group and variable", () => {
    expect(parse("Test string[ with group] and {var}")).toMatchSnapshot();
    expect(parse("Test string[ with group and {var}]")).toMatchSnapshot();
  });

  it("complex string with nested group", () => {
    expect(parse("Test string[ with group[ and {var}]]")).toMatchSnapshot();
    expect(parse("Test string[ with {group}[ and {var}]]")).toMatchSnapshot();
  });
});

describe("compose", () => {
  it("simple ast", () => {
    const ast = {
      children: [
        {
          type: "text",
          value: "Test string",
        },
      ],
      type: "root",
    };
    const values = {};
    expect(compose(ast, values)).toBe("Test string");
  });

  it("ast with variable", () => {
    const ast = {
      children: [
        {
          type: "text",
          value: "Test string with ",
        },
        {
          type: "variable",
          value: "var",
        },
      ],
      type: "root",
    };

    expect(compose(ast)).toBe("Test string with ");
    expect(compose(ast, { var: "<variable>" })).toBe(
      "Test string with <variable>"
    );
  });

  it("ast with group", () => {
    const ast = {
      children: [
        {
          type: "text",
          value: "Test string",
        },
        {
          children: [
            {
              type: "text",
              value: " with group",
            },
          ],
          type: "group",
        },
      ],
      type: "root",
    };

    expect(compose(ast)).toBe("Test string with group");
  });

  it("ast with group and variable", () => {
    const ast = {
      children: [
        {
          type: "text",
          value: "Test string",
        },
        {
          children: [
            {
              type: "text",
              value: " with group and ",
            },
            {
              type: "variable",
              value: "var",
            },
          ],
          type: "group",
        },
      ],
      type: "root",
    };

    expect(compose(ast)).toBe("Test string");
    expect(compose(ast, { var: "<variable>" })).toBe(
      "Test string with group and <variable>"
    );
  });

  it("ast with nested group", () => {
    const ast = {
      children: [
        {
          type: "text",
          value: "Test string",
        },
        {
          children: [
            {
              type: "text",
              value: " with ",
            },
            {
              type: "variable",
              value: "group",
            },
            {
              children: [
                {
                  type: "text",
                  value: " and ",
                },
                {
                  type: "variable",
                  value: "var",
                },
              ],
              type: "group",
            },
          ],
          type: "group",
        },
      ],
      type: "root",
    };

    expect(compose(ast)).toBe("Test string");
    expect(compose(ast, { group: "<group>" })).toBe("Test string with <group>");
    expect(compose(ast, { var: "<variable>" })).toBe("Test string");
    expect(compose(ast, { group: "<group>", var: "<variable>" })).toBe(
      "Test string with <group> and <variable>"
    );
  });
});
