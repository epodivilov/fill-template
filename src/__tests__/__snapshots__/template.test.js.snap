// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse complex string with nested group 1`] = `
Object {
  "children": Array [
    Object {
      "type": "text",
      "value": "Test string",
    },
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": " with group",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": " and ",
            },
            Object {
              "type": "variable",
              "value": "var",
            },
          ],
          "type": "group",
        },
      ],
      "type": "group",
    },
  ],
  "type": "root",
}
`;

exports[`parse complex string with nested group 2`] = `
Object {
  "children": Array [
    Object {
      "type": "text",
      "value": "Test string",
    },
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": " with ",
        },
        Object {
          "type": "variable",
          "value": "group",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": " and ",
            },
            Object {
              "type": "variable",
              "value": "var",
            },
          ],
          "type": "group",
        },
      ],
      "type": "group",
    },
  ],
  "type": "root",
}
`;

exports[`parse simple string 1`] = `
Object {
  "children": Array [
    Object {
      "type": "text",
      "value": "Test string",
    },
  ],
  "type": "root",
}
`;

exports[`parse string with group 1`] = `
Object {
  "children": Array [
    Object {
      "type": "text",
      "value": "Test string",
    },
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": " with group",
        },
      ],
      "type": "group",
    },
  ],
  "type": "root",
}
`;

exports[`parse string with group 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "Test string ",
        },
      ],
      "type": "group",
    },
    Object {
      "type": "text",
      "value": "with group",
    },
  ],
  "type": "root",
}
`;

exports[`parse string with group and variable 1`] = `
Object {
  "children": Array [
    Object {
      "type": "text",
      "value": "Test string",
    },
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": " with group",
        },
      ],
      "type": "group",
    },
    Object {
      "type": "text",
      "value": " and ",
    },
    Object {
      "type": "variable",
      "value": "var",
    },
  ],
  "type": "root",
}
`;

exports[`parse string with group and variable 2`] = `
Object {
  "children": Array [
    Object {
      "type": "text",
      "value": "Test string",
    },
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": " with group and ",
        },
        Object {
          "type": "variable",
          "value": "var",
        },
      ],
      "type": "group",
    },
  ],
  "type": "root",
}
`;

exports[`parse string with variable 1`] = `
Object {
  "children": Array [
    Object {
      "type": "text",
      "value": "Test string with ",
    },
    Object {
      "type": "variable",
      "value": "var",
    },
  ],
  "type": "root",
}
`;
